CRM/Utils/Hook.php

```diff
@@ -298,10 +340,8 @@ public function requireCiviModules(&$moduleList) {
    */
   public static function pre($op, $objectName, $id, &$params) {
     $event = new \Civi\Core\Event\PreEvent($op, $objectName, $id, $params);
-    \Civi::service('dispatcher')->dispatch("hook_civicrm_pre", $event);
-    \Civi::service('dispatcher')->dispatch("hook_civicrm_pre::$objectName", $event);
```

Comment:
Good question. This change come from c73e3098.

The old lines were part of a quick-and-dirty attempt to support hook-events in core. (You might refer to them as dual-emit because they were emitted once through Symfony `EventDispatcher` and once through CMS.) However, that approach required a bunch of boilerplate: you write a new PHP file (`Civi/Core/Event/PreEvent.php`) for *every hook*, and then patch *every hook* to instantiate+dispatch it.

In the revised code, lines 333-335 are redundant. If you call `dispatch('hook_foo', ...)`, then the event is dispatched everywhere. You don't need extra classes because all hooks can use `GenericHookEvent`.

There were 3 or 4 hooks which had the old boilerplate classes. I took an approach which felt more conservative -- the boilerplate classes are still there, but they now extend `GenericHookEvent`.

In retrospect, it would have been fair-game to entirely remove those boilerplate classes. Grepping universe, it appears that `PreEvent` (etal) have been internal to `civicrm-core`. But I don't mind the above because it's technically better refactoring technique.